balancer/balanceraggregator/balanceraggregator.go

```diff
@@ -0,0 +1,265 @@
+// processes the new update.
+func (ba *BalancerAggregator) UpdateClientConnState(state balancer.ClientConnState) error {
+	if len(state.ResolverState.Endpoints) == 0 {
+		return errors.New("endpoints list is empty")
+	}
+
+	// Check/return early if any endpoints have no addresses.
+	// TODO: eventually make this configurable.
+	for _, endpoint := range state.ResolverState.Endpoints {
+		if len(endpoint.Addresses) == 0 {
+			return errors.New("endpoint has empty addresses")
+		}
+	}
+	endpointSet := resolver.NewEndpointMap()
+	ba.inhibitChildUpdates = true
+	defer func() {
+		ba.inhibitChildUpdates = false
+		ba.updateState()
+	}()
+	var ret error
+	// Update/Create new children.
+	for _, endpoint := range state.ResolverState.Endpoints {
+		endpointSet.Set(endpoint, nil)
+		var bal *balancerWrapper
+		if child, ok := ba.children.Get(endpoint); ok {
+			bal = child.(*balancerWrapper)
+		} else {
+			bal = &balancerWrapper{
+				childState: ChildState{Endpoint: endpoint},
+				ClientConn: ba.cc,
+				ba:         ba,
+			}
+			bal.Balancer = gracefulswitch.NewBalancer(bal, ba.bOpts)
+			ba.children.Set(endpoint, bal)
+		}
+		if err := bal.UpdateClientConnState(balancer.ClientConnState{
+			BalancerConfig: state.BalancerConfig,
+			ResolverState: resolver.State{
+				Endpoints:  []resolver.Endpoint{endpoint},
+				Attributes: state.ResolverState.Attributes,
+			},
+		}); err != nil && ret == nil {
+			// Return first error found, and always commit full processing of
+			// updating children. If desired to process more specific errors
+			// across all endpoints, caller should make these specific
+			// validations, this is a current limitation for simplicities sake.
+			ret = err
+		}
+	}
+	// Delete old children that are no longer present.
+	for _, e := range ba.children.Keys() {
+		child, _ := ba.children.Get(e)
+		bal := child.(balancer.Balancer)
+		if _, ok := endpointSet.Get(e); !ok {
+			bal.Close()
+			ba.children.Delete(e)
+		}
+	}
+	return ret
+}
+
+// ResolverError forwards the resolver error to all of the BalancerAggregator's
+// children, sends a single synchronous update of the childStates at the end of
+// the ResolverError operation.
+func (ba *BalancerAggregator) ResolverError(err error) {
+	ba.inhibitChildUpdates = true
+	defer func() {
+		ba.inhibitChildUpdates = false
+		ba.updateState()
+	}()
+	for _, child := range ba.children.Values() {
+		bal := child.(balancer.Balancer)
+		bal.ResolverError(err)
+	}
+}
+
+func (ba *BalancerAggregator) UpdateSubConnState(sc balancer.SubConn, state balancer.SubConnState) {
+	// UpdateSubConnState is deprecated.
+}
+
+func (ba *BalancerAggregator) Close() {
+	for _, child := range ba.children.Values() {
+		bal := child.(balancer.Balancer)
+		bal.Close()
+	}
+}
+
+// updateState updates this components state. It sends the aggregated state, and
+// a picker with round robin behavior with all the child states present if
+// needed.
+func (ba *BalancerAggregator) updateState() {
```

Comment:
Added a lock on the child, and hold it in UpdateState() write, and updateState() read. I think no deadlocks occur, and I think this is correct.

The reason I did it per child is to avoid unnecessary blocking since the read is scoped to a child. This works if the goal is eventual consistency thinking through the scenarios.

Say you get an UpdateCCS for endpoint A, B, writes current state, calls update state (lock not held) at the end. endpoint B sends a new update before UpdateState() on parent. First Update will be endpointAState, and endpointBNewState, and then updateState() from endpointB update will trigger a resend. An extra duplicate update I think from previous experience is acceptable, and not technically incorrect. The initial singular updateState() from UpdateCCS() could get a new update from a child after calling updateState(), but this feels fine.

In the case of getting numerous interleaved UpdateState() from numerous, the updateState() *after* unlocking in UpdateState() makes the last update trigger the updateState() operation which will eventually reflect the most recent update states to the parent.

Let me know what you think.

Reply:
There's far too many "update"s and "state"s in that for me to make sense of it, sorry.

I'm pretty sure we want to prevent multiple concurrent calls to the parent LB policy's `UpdateState` method.  Our channel implementation handles it just fine, but ideally we can make this serialize those calls to simplify the implementation of the parent LB policy.

Reply:
...in which case we can probably just use the same lock to guard `inhibitChildUpdates` instead of worrying about making it atomic...